### ISSUE TITLE
Error writing Galera Config.  Permission Denied[Bug] 

### ISSUE THREAD

**Issue by anfechtung** [mariadb-operator/mariadb-operator#171]
**Describe the bug**
I am attempting to follow the basic documentation for deploying a galera cluster using the operator.

From the main branch of the mariadb-operator

```
DOCKER-SHELL1:/tmp/mariadb-operator$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
```

I create the config manifests:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ kubectl apply -f examples/manifests/config
configmap/mariadb created
configmap/mariadb-my-cnf created
secret/mariadb created
secret/user created
```

Followed by the galera cluster manifest:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ kubectl apply -f examples/manifests/mariadb_v1alpha1_mariadb_galera.yaml
mariadb.mariadb.mmontes.io/mariadb-galera created
```

The cluster fails to initialize:
```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get po
NAME                                        READY   STATUS                  RESTARTS        AGE
mariadb-galera-0                            0/2     Init:CrashLoopBackOff   6 (4m15s ago)   10m
mariadb-galera-1                            0/2     Init:CrashLoopBackOff   6 (4m17s ago)   10m
mariadb-galera-2                            0/2     Init:CrashLoopBackOff   6 (4m5s ago)    10m
mariadb-operator-7bcf6d44f-9nnks            1/1     Running                 0               3h15m
mariadb-operator-webhook-6d7669464d-84hgm   1/1     Running                 0           
```

With the following init errors

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs mariadb-galera-0 init
{"level":"info","ts":1690491722.3708434,"msg":"Starting init"}
{"level":"info","ts":1690491722.3937182,"msg":"Configuring Galera"}
{"level":"error","ts":1690491722.393788,"msg":"Error writing Galera config","error":"open /etc/mysql/mariadb.conf.d/0-galera.cnf: permission denied","stacktrace":"runtime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
DOCKER-SHELL1:/tmp/mariadb-operator$ 
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs mariadb-galera-0 init --previous
{"level":"info","ts":1690491722.3708434,"msg":"Starting init"}
{"level":"info","ts":1690491722.3937182,"msg":"Configuring Galera"}
{"level":"error","ts":1690491722.393788,"msg":"Error writing Galera config","error":"open /etc/mysql/mariadb.conf.d/0-galera.cnf: permission denied","stacktrace":"runtime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
DOCKER-SHELL1:/tmp/mariadb-operator$ 
'''


**Environment details**:
- Kubernetes version: v1.25.5
- mariadb-operator version: mariadb-operator-0.19.0 v0.0.19 
- Install method: helm
- Install flavour: minimal

**Comment by mmontes11** [comment]
Hey @anfechtung ! Thanks for bringing this up.

The [init container](https://github.com/mariadb-operator/init) is attempting to populate the galera configuration file under `/etc/mysql/mariadb.conf.d` which is shared with the MariaDB container.

Problem is that, for some reason, the `initContainer` does not have permission to write in the PVC. Some users reported this already, please take a look at [this](https://github.com/mariadb-operator/mariadb-operator/issues/155#issuecomment-1637151587). Which Kubernetes distribution and storage provider are you using in order to identify this?

To mitigate the issue, you could provide the following `spec.podSecurityContext`: 
```yaml
apiVersion: mariadb.mmontes.io/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
spec:
  ...
  galera:
    enabled: true
  podSecurityContext:
    runAsUser: 0
```
Could you give it a try and let me know how it goes? Thanks!

**Comment by jonizen** [comment]
I had the exact same issue but running it on microk8s and longhorn for storage, I can try it out and report back :)

**Comment by jonizen** [comment]
> Hey @anfechtung ! Thanks for bringing this up.
> 
> The [init container](https://github.com/mariadb-operator/init) is attempting to populate the galera configuration file under `/etc/mysql/mariadb.conf.d` which is shared with the MariaDB container.
> 
> Problem is that, for some reason, the `initContainer` does not have permission to write in the PVC. Some users reported this already, please take a look at [this](https://github.com/mariadb-operator/mariadb-operator/issues/155#issuecomment-1637151587). Which Kubernetes distribution and storage provider are you using in order to identify this?
> 
> To mitigate the issue, you could provide the following `spec.podSecurityContext`:
> 
> ```yaml
> apiVersion: mariadb.mmontes.io/v1alpha1
> kind: MariaDB
> metadata:
>   name: mariadb-galera
> spec:
>   ...
>   galera:
>     enabled: true
>   podSecurityContext:
>     runAsUser: 0
> ```
> 
> Could you give it a try and let me know how it goes? Thanks!

So I bumped the init to v0.0.5 and nothing else and it did not work as expected, then added podSecurityContext and it works!

`{"level":"info","ts":1690534824.5494747,"msg":"Starting init"}
{"level":"info","ts":1690534824.6221566,"msg":"Configuring Galera"}
{"level":"info","ts":1690534824.6227636,"msg":"Configuring bootstrap"}
{"level":"info","ts":1690534824.622967,"msg":"Init done"}`

**Comment by jonizen** [comment]
I have tested it out, connected to the database and querying some data and works. Thanks for all the support! @mmontes11

**Comment by mmontes11** [comment]
Thanks a lot @jonizen ! I'm glad that it finally works. I'm not a longhorn expert, do you happen to know why it requires a different set of permissions? Maybe writing to the PVC can only be performed by specific users? I would like to understand this as it is not the first time that happens.

@anfechtung please [give it a try](https://github.com/mariadb-operator/mariadb-operator/issues/171#issuecomment-1655066329), just to confirm so we can close this one.

**Comment by anfechtung** [comment]
I checked the MariaDB spec, and it already was using the `v0.0.5` version of the init container.

I trued the run as approach

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get mariadb mariadb-galera -o yaml | grep -C1 "runAsUser"
--
  podSecurityContext:
    runAsUser: 0
  port: 3306
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

And I am now getting a different configuration error at a later stage:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get po
NAME                                        READY   STATUS             RESTARTS       AGE
mariadb-galera-0                            1/2     CrashLoopBackOff   12 (39s ago)   37m
mariadb-galera-1                            0/2     Init:0/1           0              37m
mariadb-galera-2                            0/2     Init:0/1           0              37m
mariadb-operator-7bcf6d44f-9nnks            1/1     Running            0              22h
mariadb-operator-webhook-6d7669464d-84hgm   1/1     Running            0              22h
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs mariadb-galera-0
Defaulted container "mariadb" out of: mariadb, agent, init (init)
2023-07-28 15:59:56+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.0.2+maria~ubu2204 started.
2023-07-28 15:59:56+00:00 [ERROR] [Entrypoint]: mariadbd failed while attempting to check config
        command was: mariadbd --verbose --help
        Can't initialize timers
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

For reference the pvcs are provisioned by rook-ceph os RWO volumes with an external RBD backend.

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get pvc
NAME                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
galera-mariadb-galera-0    Bound    pvc-28bdba3d-fb7c-4f48-85c4-4335b0f2fc48   50Mi       RWO            ceph-rbd       45m
galera-mariadb-galera-1    Bound    pvc-a5c94ecb-4e79-487a-b44b-80b50457faf4   50Mi       RWO            ceph-rbd       45m
galera-mariadb-galera-2    Bound    pvc-f2c4994d-af5b-4541-b00f-a1d9e4980b3e   50Mi       RWO            ceph-rbd       45m
storage-mariadb-galera-0   Bound    pvc-7c9f3074-ba51-4c52-ad8d-7fd06414e511   1Gi        RWO            ceph-rbd       45m
storage-mariadb-galera-1   Bound    pvc-4f9be1e7-c118-4cb6-9740-8115778e463a   1Gi        RWO            ceph-rbd       45m
storage-mariadb-galera-2   Bound    pvc-647a2b52-3df6-4907-903b-b6b3c835fc1d   1Gi        RWO            ceph-rbd       45m
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

**Comment by mmontes11** [comment]
hey @anfechtung ! thanks for reporting back

> DOCKER-SHELL1:/tmp/mariadb-operator$ k logs mariadb-galera-0
Defaulted container "mariadb" out of: mariadb, agent, init (init)
2023-07-28 15:59:56+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.0.2+maria~ubu2204 started.
2023-07-28 15:59:56+00:00 [ERROR] [Entrypoint]: mariadbd failed while attempting to check config
        command was: mariadbd --verbose --help
        Can't initialize timers
        
I'm affraid these are the logs of the `mariadb` container, could you provide the ones from the `init` container?:

```bash
 kubectl logs mariadb-galera-0 -c init
```

**Comment by anfechtung** [comment]
```
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ kubectl logs mariadb-galera-0 -c init
{"level":"info","ts":1690566983.6210072,"msg":"Starting init"}
{"level":"info","ts":1690566983.651107,"msg":"Configuring Galera"}
{"level":"info","ts":1690566983.6675305,"msg":"Configuring bootstrap"}
{"level":"info","ts":1690566983.6677482,"msg":"Init done"}
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ 
```

**Comment by mmontes11** [comment]
@anfechtung interesting, I think your config file either got corrupted or is invalid, the init container seem to have worked. Could you provide the contents of `/etc/mysql/mariadb.conf.d/0-galera.cnf`?

@grooverdan regarding the MariaDB container logs, does this logs ring a bell?

**Comment by jonizen** [comment]
Did you recreate the galera cluster or just change the manifest and version for the init container? I had to recreate it, to make it work 😊


﻿

@anfechtung<https://github.com/anfechtung> interesting, I think your config file either got corrupted or is invalid, the init container seem to have worked. Could you provide the contents of /etc/mysql/mariadb.conf.d/0-galera.cnf?

@grooverdan<https://github.com/grooverdan> regarding the MariaDB container logs, does this logs ring a bell?

**Comment by jonizen** [comment]
> Thanks a lot @jonizen ! I'm glad that it finally works. I'm not a longhorn expert, do you happen to know why it requires a different set of permissions? Maybe writing to the PVC can only be performed by specific users? I would like to understand this as it's not the first time it happens.
> 
> 
> 
> @anfechtung please [give it a try](https://github.com/mariadb-operator/mariadb-operator/issues/171#issuecomment-1655066329), just to confirm so we can close this one.
> 
> 
> 
> 

I am no expert on longhorn either 😊 but know this request, talking about this. https://github.com/longhorn/longhorn/issues/1165

The request is for the possibility of have specific permissions for init. At that time longhorn did not respect security on text by do now.

And this reference to update the docs.

“Currently, Longhorn uses Bidirectional mount propagation, therefore it is allowed only in privileged containers. Ref to https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation”

https://github.com/longhorn/longhorn/issues/3549

https://github.com/longhorn/longhorn/issues/3549#issuecomment-1046494049

Not sure this is to any help 😅

**Comment by grooverdan** [comment]
"Can't initialize timers" -> https://github.com/MariaDB/mariadb-docker/issues/434

TLDR;
* update docker version
* seccomp filter of the mariadb container isn't sufficient for it to run

**Comment by anfechtung** [comment]
@grooverdan thanks for the information.  I was able to get past *that* error by allowing the pod to run in an unconfined seccomp profile:

```
apiVersion: mariadb.mmontes.io/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
  namespace: cluster
spec:
  podSecurityContext:
    runAsUser: 0
  securityContext:
    seccompProfile:
      type: Unconfined
```

Next I encountered the following error:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs -n cluster mariadb-galera-0
Defaulted container "mariadb" out of: mariadb, agent, init (init)
2023-07-31 15:30:44+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.0.2+maria~ubu2204 started.
2023-07-31 15:30:44+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
2023-07-31 15:30:44+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.0.2+maria~ubu2204 started.
2023-07-31 15:30:45+00:00 [Note] [Entrypoint]: Initializing database files
2023-07-31 15:30:45 0 [ERROR] mariadbd: Error writing file './ddl_recovery.log' (Errcode: 28 "No space left on device")
2023-07-31 15:30:45 0 [ERROR] Aborting
2023-07-31 15:30:45 0 [ERROR] mariadbd: Error on close of '(null)' (Errcode: 9 "Bad file descriptor")
```

I am working on debugging that issue now.

**Comment by anfechtung** [comment]
It looks like bumping up pvc to 500Mi worked.  I *think* you can close the issue.  However you may want to change the default size, or update the documentation for galera clusters appropriately.

**Comment by anfechtung** [comment]
However, before you close the ticket, I have another question/issue.  One of the reasons I am looking at using the mariadb operator is the cluster recovery feature: 

Multi-master HA via [Galera](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/GALERA.md)✨. Automatic Galera cluster [recovery](https://galeracluster.com/library/documentation/crash-recovery.html) ✨

Once I got my cluster running I simulated node failure.  Basically force stopped and restarted the underlying vms.  After the nodes came back online I see the mariadb pods are stuck in a crash loop:

```
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ k get po -n cluster
NAME               READY   STATUS             RESTARTS         AGE
mariadb-galera-0   1/2     CrashLoopBackOff   12 (2m54s ago)   56m
mariadb-galera-1   1/2     CrashLoopBackOff   11 (110s ago)    56m
mariadb-galera-2   1/2     CrashLoopBackOff   11 (111s ago)    56m
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ 
```

with the following error on node 0:

```
2023-07-31 21:45:00 0 [Note] WSREP: Passing config to GCS: base_dir = /var/lib/mysql/; base_host = mariadb-galera-0.mariadb-galera-internal.cluster.svc.cluster.local; base_port = 4567; cert.log_conflicts = no; cert.optimistic_pa = yes; debug = no; evs.auto_evict = 0; evs.delay_margin = PT1S; evs.delayed_keep_period = PT30S; evs.inactive_check_period = PT0.5S; evs.inactive_timeout = PT15S; evs.join_retrans_period = PT1S; evs.max_install_timeouts = 3; evs.send_window = 4; evs.stats_report_period = PT1M; evs.suspect_timeout = PT5S; evs.user_send_window = 2; evs.view_forget_timeout = PT24H; gcache.dir = /var/lib/mysql/; gcache.keep_pages_size = 0; gcache.keep_plaintext_size = 128M; gcache.mem_size = 0; gcache.name = galera.cache; gcache.page_size = 128M; gcache.recover = yes; gcache.size = 128M; gcomm.thread_prio = ; gcs.fc_debug = 0; gcs.fc_factor = 1.0; gcs.fc_limit = 16; gcs.fc_master_slave = no; gcs.fc_single_primary = no; gcs.max_packet_size = 64500; gcs.max_throttle = 0.25; gcs.recv_q_hard_limit = 9223372036854775807; gcs.recv_q_soft_
2023-07-31 21:45:00 0 [Note] WSREP: Start replication
2023-07-31 21:45:00 0 [Note] WSREP: Connecting with bootstrap option: 1
2023-07-31 21:45:00 0 [Note] WSREP: Setting GCS initial position to 00000000-0000-0000-0000-000000000000:-1
2023-07-31 21:45:00 0 [ERROR] WSREP: It may not be safe to bootstrap the cluster from this node. It was not the last one to leave the cluster and may not contain all the updates. To force cluster bootstrap with this node, edit the grastate.dat file manually and set safe_to_bootstrap to 1 .
2023-07-31 21:45:00 0 [ERROR] WSREP: wsrep::connect(gcomm://mariadb-galera-0.mariadb-galera-internal.cluster.svc.cluster.local,mariadb-galera-1.mariadb-galera-internal.cluster.svc.cluster.local,mariadb-galera-2.mariadb-galera-internal.cluster.svc.cluster.local) failed: 7
2023-07-31 21:45:00 0 [ERROR] Aborting
```

Are there additional steps or components required for the automated recovery?

**Comment by anfechtung** [comment]
Okay, I recreated the full cluster:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k create ns cluster
namespace/cluster created
DOCKER-SHELL1:/tmp/mariadb-operator$ k apply -f examples/manifests/config/
configmap/mariadb created
configmap/mariadb-my-cnf created
secret/mariadb created
secret/user created
DOCKER-SHELL1:/tmp/mariadb-operator$  k apply -f examples/manifests/mariadb_v1alpha1_mariadb_galera_minimal.yaml
mariadb.mariadb.mmontes.io/mariadb-galera created
DOCKER-SHELL1:/tmp/mariadb-operator$ k get po -n cluster
NAME               READY   STATUS    RESTARTS   AGE
mariadb-galera-0   2/2     Running   0          3m8s
mariadb-galera-1   2/2     Running   0          3m7s
mariadb-galera-2   1/2     Running   0          3m7s
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

I then forced a vm shutdown/restart to simulate a complete cluster outage.

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get po -n cluster
NAME               READY   STATUS             RESTARTS        AGE
mariadb-galera-0   1/2     CrashLoopBackOff   8 (3m48s ago)   29m
mariadb-galera-1   1/2     CrashLoopBackOff   7 (2m40s ago)   29m
mariadb-galera-2   1/2     CrashLoopBackOff   8 (12s ago)     29m
DOCKER-SHELL1:/tmp/mariadb-operator$ kubectl get mariadb mariadb-galera -o jsonpath="{.status}" -n cluster
{"conditions":[{"lastTransitionTime":"2023-08-01T16:14:47Z","message":"Not ready","reason":"StatefulSetNotReady","status":"False","type":"Ready"}],"currentPrimary":"All"}
DOCKER-SHELL1:/tmp/mariadb-operator$ 
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs mariadb-operator-7bcf6d44f-9nnks --tail 15
{"level":"info","ts":1690907395.1857297,"msg":"Starting EventSource","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","source":"kind source: *v1.ClusterRoleBinding"}
{"level":"info","ts":1690907395.185738,"msg":"Starting EventSource","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","source":"kind source: *v1.ServiceMonitor"}
{"level":"info","ts":1690907395.1857443,"msg":"Starting Controller","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB"}
{"level":"info","ts":1690907395.1857283,"msg":"Starting EventSource","controller":"pod","controllerGroup":"","controllerKind":"Pod","source":"kind source: *v1.Pod"}
{"level":"info","ts":1690907395.1864786,"msg":"Starting Controller","controller":"pod","controllerGroup":"","controllerKind":"Pod"}
{"level":"info","ts":1690907395.957341,"msg":"Starting workers","controller":"statefulset","controllerGroup":"apps","controllerKind":"StatefulSet","worker count":1}
{"level":"info","ts":1690907395.9950216,"msg":"Starting workers","controller":"pod","controllerGroup":"","controllerKind":"Pod","worker count":1}
{"level":"info","ts":1690907395.9961407,"msg":"Starting workers","controller":"grant","controllerGroup":"mariadb.mmontes.io","controllerKind":"Grant","worker count":1}
{"level":"info","ts":1690907395.9961412,"msg":"Starting workers","controller":"connection","controllerGroup":"mariadb.mmontes.io","controllerKind":"Connection","worker count":1}
{"level":"info","ts":1690907396.0182736,"msg":"Starting workers","controller":"user","controllerGroup":"mariadb.mmontes.io","controllerKind":"User","worker count":1}
{"level":"info","ts":1690907396.0183854,"msg":"Starting workers","controller":"database","controllerGroup":"mariadb.mmontes.io","controllerKind":"Database","worker count":1}
{"level":"info","ts":1690907396.018428,"msg":"Starting workers","controller":"backup","controllerGroup":"mariadb.mmontes.io","controllerKind":"Backup","worker count":1}
{"level":"info","ts":1690907396.018482,"msg":"Starting workers","controller":"sqljob","controllerGroup":"mariadb.mmontes.io","controllerKind":"SqlJob","worker count":1}
{"level":"info","ts":1690907396.018512,"msg":"Starting workers","controller":"restore","controllerGroup":"mariadb.mmontes.io","controllerKind":"Restore","worker count":1}
{"level":"info","ts":1690907396.0477326,"msg":"Starting workers","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","worker count":1}
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

There was nothing germane in the event logs.

**Comment by anfechtung** [comment]
Okay, so I keep finding things.  

If I change *only* the password in the mariadb-secret.yaml:

```
apiVersion: v1
kind: Secret
metadata:
  name: mariadb
  namespace: cluster
data:
  password: aGVsbG93b3JsZAo=
  root-password: aGVsbG93b3JsZAo=
```

my cluster ends up in a crash loop with:

```
DOCKER-SHELL1:/tmp/mariadb-operator$ k get po -n cluster
NAME               READY   STATUS             RESTARTS      AGE
mariadb-galera-0   1/2     CrashLoopBackOff   2 (33s ago)   71s
mariadb-galera-1   0/2     Init:0/1           0             70s
mariadb-galera-2   0/2     Init:0/1           0             69s
DOCKER-SHELL1:/tmp/mariadb-operator$ k logs -n cluster mariadb-galera-0 
Defaulted container "mariadb" out of: mariadb, agent, init (init)
2023-08-01 22:48:38+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.0.2+maria~ubu2204 started.
2023-08-01 22:48:38+00:00 [ERROR] [Entrypoint]: mariadbd failed while attempting to check config
        command was: mariadbd --verbose --help
        2023-08-01 22:48:38 0 [Warning] Could not open mysql.plugin table: "Table 'mysql.plugin' doesn't exist". Some options may be missing from the help text
2023-08-01 22:48:38 0 [ERROR] mariadbd: unknown option '--"'
DOCKER-SHELL1:/tmp/mariadb-operator$ 
```

**Comment by mmontes11** [comment]
Hey  @anfechtung! Unfotunately this doesn't help too much. We need to understand why the operator is not kicking in after all the cluster nodes being unhealthy for a period of time, defined by [spec.galera.clusterHealthyTimeout](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_mariadb_galera.yaml#L46) and defaulted to [1m](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_mariadb_galera.yaml#L46)

I would suggest setting the `debug` log level in the operator. This can be done by setting `loglevel=debug` in the helm chart:

https://github.com/mariadb-operator/mariadb-operator/blob/b112d4b2e2397fcee2c38d1f95d6673795e592ba/deploy/charts/mariadb-operator/values.yaml#L12

or manually setting the `--log-level=debug` flag in the operator `Deployment`:

https://github.com/mariadb-operator/mariadb-operator/blob/b112d4b2e2397fcee2c38d1f95d6673795e592ba/deploy/charts/mariadb-operator/templates/deployment.yaml#L52

Then, you may tear down your cluster nodes and check the operator logs. 

Also to better understand the process, it would be great to know your `spec.galera.clusterHealthyTimeout` value:
```bash
kubectl get mariadb mariadb-galera -o jsonpath="{.spec.galera.clusterHealthyTimeou}"
```

**Comment by mmontes11** [comment]
> First off, do you have any thoughts on the password issue I noted above? Should I open a separate Issue for that?

Sorry forgot about this point. We are relying on the [MARIADB_ROOT_PASSWORD](https://mariadb.com/kb/en/mariadb-server-docker-official-image-environment-variables/#mariadb_root_password_hash-mariadb_root_password-mysql_root_password) for setting the root password initially, but AFAIK the MariaDB oficial image only uses it when initializing the instance and it won't be overriden. In any case, I don't think it is related to the current issue, could you create a separate one please?

> Error reconciling Galera: error disabling bootstrap in Pod 0: unauthorized

Right, this is an RBAC issue that other users have already experienced. Could you please check this [comment](https://github.com/mariadb-operator/mariadb-operator/issues/159#issuecomment-1643430626) please? Basically an extra `ClusterRoleBinding` to inherit the `system:auth-delegator` permissions is required for both the operator and the MariaDB instances. The operator CRB is created by helm whereas the MariaDB is created on the flight by the operator.

I'm wondering if the `system:auth-delegator` `ClusterRole` has limitations with particular Kubernetes versions, will check this on my side.

**Comment by anfechtung** [comment]
I feel like I am very close now, but still something is missing.  I recreated my cluster, and added a cluster role binding.  After reboot two of the nodes come up cleanly.  However the third node fails.

```
DOCKER-SHELL1:/tmp/fb-mariadb$ k get po -n cluster
NAME               READY   STATUS    RESTARTS     AGE
mariadb-galera-0   2/2     Running   0            10m
mariadb-galera-1   2/2     Running   0            10m
mariadb-galera-2   1/2     Running   1 (6s ago)   2m18s
DOCKER-SHELL1:/tmp/fb-mariadb$ 
```

The error:

```
2023-08-02 21:42:10 2 [Note] WSREP: Server status change joiner -> connected
2023-08-02 21:42:10 2 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2023-08-02 21:42:10 2 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2023-08-02 21:42:10 2 [Note] WSREP: ================================================
View:
  id: 00000000-0000-0000-0000-000000000000:14
  status: non-primary
  protocol_version: 4
  capabilities: MULTI-MASTER, CERTIFICATION, PARALLEL_APPLYING, REPLAY, ISOLATION, PAUSE, CAUSAL_READ, INCREMENTAL_WS, UNORDERED, PREORDERED, STREAMING, NBO
  final: yes
  own_index: -1
  members(0):
=================================================
2023-08-02 21:42:10 2 [Note] WSREP: Non-primary view
2023-08-02 21:42:10 2 [Note] WSREP: Server status change connected -> disconnected
2023-08-02 21:42:10 2 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2023-08-02 21:42:10 2 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2023-08-02 21:42:10 2 [Note] WSREP: Applier thread exiting ret: 0 thd: 2
2023-08-02 21:42:10 2 [Warning] Aborted connection 2 to db: 'unconnected' user: 'unauthenticated' host: '' (This connection closed normally without authentication)
```

From the operator logs, the sync process appears to timeout and the pod is terminated:

```
{"level":"error","ts":1691012533.389296,"logger":"galera.recovery.pod","msg":"Timeout waiting for Pod to be Synced. Deleting Pod
","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","mariaDB":{"name":"mariadb-galera","n
amespace":"cluster"},"namespace":"cluster","name":"mariadb-galera","reconcileID":"e708511a-7716-4e95-b028-906da53475e3","pod":"m
ariadb-galera-2","timeout":"3m0s","error":"timed out waiting for the condition","stacktrace":"github.com/mariadb-operator/mariad
b-operator/pkg/controller/galera.(*GaleraReconciler).recoverPods.func1\n\t/app/pkg/controller/galera/recovery.go:115"}
{"level":"debug","ts":1691012533.3920653,"logger":"events","msg":"Warning","object":{"kind":"MariaDB","namespace":"cluster","nam
e":"mariadb-galera","uid":"31e3b80e-a90d-42f8-a355-83e053adacff","apiVersion":"mariadb.mmontes.io/v1alpha1","resourceVersion":"4
9391"},"reason":"GaleraPodSyncTimeout","message":"Timeout waiting for Pod 'mariadb-galera-2' to be Synced"}
```

I am going to try futzing with the galera recovery timeouts and see if that resolves the issue.

**Comment by anfechtung** [comment]
It looks llike upping the timeouts fixed the issue.  The operator successfully recovered the cluster post crash:

```
{"level":"info","ts":1691016927.1725614,"logger":"galera.health","msg":"Galera cluster is not is not healthy","controller":"statefulset","controllerGroup":"apps","controllerKind":"StatefulSet","statefulSet":{"name":"mariadb-galera","namespace":"cluster"},"namespace":"cluster","name":"mariadb-galera","reconcileID":"17ca5044-013d-45bb-b609-dc52bbaaf393"}
{"level":"debug","ts":1691016927.1745093,"logger":"events","msg":"Warning","object":{"kind":"MariaDB","namespace":"cluster","name":"mariadb-galera","uid":"a73f0e94-b5ba-47c5-b5ea-362d5c62a47f","apiVersion":"mariadb.mmontes.io/v1alpha1","resourceVersion":"97014"},"reason":"GaleraClusterNotHealthy","message":"Galera cluster is not healthy"}
{"level":"debug","ts":1691016927.4950378,"logger":"galera","msg":"Disabling Galera bootstrap","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","mariaDB":{"name":"mariadb-galera","namespace":"cluster"},"namespace":"cluster","name":"mariadb-galera","reconcileID":"b234687f-0f80-4c4e-938c-48cf54a3c928"}
{"level":"info","ts":1691016927.7260163,"logger":"galera","msg":"Galera cluster is healthy","controller":"mariadb","controllerGroup":"mariadb.mmontes.io","controllerKind":"MariaDB","mariaDB":{"name":"mariadb-galera","namespace":"cluster"},"namespace":"cluster","name":"mariadb-galera","reconcileID":"b234687f-0f80-4c4e-938c-48cf54a3c928"}
{"level":"debug","ts":1691016927.728437,"logger":"events","msg":"Normal","object":{"kind":"MariaDB","namespace":"cluster","name":"mariadb-galera","uid":"a73f0e94-b5ba-47c5-b5ea-362d5c62a47f","apiVersion":"mariadb.mmontes.io/v1alpha1","resourceVersion":"107284"},"reason":"GaleraClusterHealthy","message":"Galera cluster is healthy"}
```

For reference I set these timeout values:

```
  galera:
    enabled: true
    recovery:
      clusterBootstrapTimeout: 10m0s
      clusterHealthyTimeout: 5m0s
      enabled: true
      podRecoveryTimeout: 10m0s
      podSyncTimeout: 10m0s
```

Thanks for all of your help!

**Comment by mmontes11** [comment]
Hey @anfechtung ! I'm glad that it finally worked.

> {"level":"debug","ts":1691012533.3920653,"logger":"events","msg":"Warning","object":{"kind":"MariaDB","namespace":"cluster","nam
e":"mariadb-galera","uid":"31e3b80e-a90d-42f8-a355-83e053adacff","apiVersion":"mariadb.mmontes.io/v1alpha1","resourceVersion":"4
9391"},"reason":"GaleraPodSyncTimeout","message":"Timeout waiting for Pod 'mariadb-galera-2' to be Synced"}

We have 3m as default [PodRecoveryTimeout](https://github.com/mariadb-operator/mariadb-operator/blob/main/api/v1alpha1/mariadb_galera_types.go#L229) which seems a bit low. Will increase it as you suggested along with the others.

Also, I will be adding a **troubleshooting** section at the end of the [GALERA.md](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/GALERA.md) so we can see all the problems and its remediations in a glance.

Closing! don't hesitate in opening another issue if you need to.

**Comment by anfechtung** [comment]
I seem to be having intermittent issues when testing cluster recovery.  Sometimes the cluster recovers.  However, sometimes I end up with pods crashing during init:

```
mariadb-galera-0                      2/2     Running                 9 (7m34s ago)    62m
mariadb-galera-1                      0/2     Init:CrashLoopBackOff   10 (4m27s ago)   62m
mariadb-galera-2                      1/2     Running                 4 (117s ago)     6m57s
```

with the error:

```
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ k logs -n frostbite mariadb-galera-1 init
{"level":"info","ts":1691092514.9077911,"msg":"Starting init"}
{"level":"error","ts":1691092514.9220397,"msg":"Error getting MariaDB","error":"error requesting 'mariadb-galera' MariaDB in namespace 'cluster': the server has asked for the client to provide credentials","stacktrace":"runtime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
DOCKER-SHELL1:/data3/jonathan/workspace/ansible-frostbite$ 

```

**Comment by mmontes11** [comment]
one more thing @anfechtung, if you happen to experiment this issue again
```bash
Error getting MariaDB","error":"error requesting 'mariadb-galera' MariaDB in namespace 'cluster': the server has asked for the client to provide credentials
```
you may try to execute this command impersonating the `mariadb-galera` serviceaccount, this is basically the same request as the one done by the `init` container:

```bash
kubectl get mariadb mariadb-galera  --as=system:serviceaccount:default:mariadb-galera
NAME             READY   STATUS    PRIMARY POD   AGE
mariadb-galera   True    Running   All           22h
```
